onnxruntime/core/framework/session_state.cc

```diff
@@ -234,36 +235,34 @@ void SessionState::CleanInitializedTensorsFromGraph() {
   graph_.CleanAllInitializedTensors();
 }
 
-Status SessionState::PrepackInitializedConstantTensors() {
-  // calculate the use count of each value
-  std::unordered_map<std::string, size_t> node_arg_use_count;
-  for (const auto& node : GetGraphViewer().Nodes()) {
-    node.ForEachDef([&](const onnxruntime::NodeArg& node_arg, bool is_input) {
-      if (is_input) {
-        node_arg_use_count[node_arg.Name()]++;
-      }
-    });
-  }
-
+Status SessionState::PrepackConstantInitializedTensors(std::unordered_map<std::string, size_t>& constant_initializers_use_count) {
   for (auto& node : GetGraphViewer().Nodes()) {
     auto kernel = GetMutableKernel(node.Index());
     int input_idx = 0;
     for (auto& input_def : node.InputDefs()) {
       if (input_def->Exists()) {
         const std::string& input_name = input_def->Name();
-        int ort_value_idx;
-        ORT_RETURN_IF_ERROR(ort_value_name_idx_map_.GetIdx(input_name, ort_value_idx));
-        if (constant_initialized_tensors_.count(ort_value_idx) &&
-            constant_initialized_tensors_[ort_value_idx].IsTensor()) {
-          bool is_packed = false;
-          const Tensor& const_initialized_tensor = constant_initialized_tensors_[ort_value_idx].Get<Tensor>();
-          ORT_RETURN_IF_ERROR(kernel->PrePack(const_initialized_tensor, input_idx, is_packed));
-          if (is_packed && node_arg_use_count.count(input_name) && --node_arg_use_count[input_name] == 0) {
-            // release the constant intialized tensor
-            initialized_tensors_.erase(ort_value_idx);
-            constant_initialized_tensors_.erase(ort_value_idx);
+        SessionState* st = this;
+        // subgraph can use the value from outer scope,
+        // so it needs to check if current node uses const initilized tensor from current and outer graphs
+        do {
+          int ort_value_idx;
+          if (st->GetOrtValueNameIdxMap().GetIdx(input_name, ort_value_idx).IsOK()) {
+            std::unordered_map<int, OrtValue>& constant_initialized_tensors = st->constant_initialized_tensors_;
+            if (constant_initialized_tensors.count(ort_value_idx)) {
+              bool is_packed = false;
+              const Tensor& const_initialized_tensor = constant_initialized_tensors[ort_value_idx].Get<Tensor>();
+              ORT_RETURN_IF_ERROR(kernel->PrePack(const_initialized_tensor, input_idx, is_packed));
+              if (is_packed && constant_initializers_use_count.count(input_name) && --constant_initializers_use_count[input_name] == 0) {
+                // release the constant intialized tensor
+                st->initialized_tensors_.erase(ort_value_idx);
+                constant_initialized_tensors.erase(ort_value_idx);
+              }
+              break;
+            }
           }
```

Comment:
If you found an entry at this level of the graph, but it's not a constant initializer, shouldn't you stop at that point as that value will shadow any outer scope constant initializer? 

i.e. the `break;` should be outside of the `if (constant_initialized_tensors.count(ort_value_idx)) {`

Reply:
Yes, it should stop. Moving break one level out can not handle the case correctly. st->GetOrtValueNameIdxMap().GetIdx(input_name, ort_value_idx).IsOK() is expected to be true for subgraph even the value comes from outer scope. I need an extra special check for subgraph.

Reply:
Graph::IsOuterScopeValue

---
In reply to: [506592788](https://github.com/microsoft/onnxruntime/pull/5433#discussion_r506592788) [](ancestors = 506592788)